sql/expression/function/srid.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2020-2021 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package function
+
+import (
+	"fmt"
+	"strings"
+
+	"gopkg.in/src-d/go-errors.v1"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+)
+
+// SRID is a function that returns SRID of Geometry object or returns a new object with altered SRID.
+type SRID struct {
+	expression.NaryExpression
+}
+
+var _ sql.FunctionExpression = (*SRID)(nil)
+
+var ErrInvalidSRID = errors.NewKind("There's no spatial reference with SRID %d")
+
+// NewSRID creates a new STX expression.
+func NewSRID(args ...sql.Expression) (sql.Expression, error) {
+	if len(args) != 1 && len(args) != 2 {
+		return nil, sql.ErrInvalidArgumentNumber.New("ST_SRID", "1 or 2", len(args))
+	}
+	return &SRID{expression.NaryExpression{ChildExpressions: args}}, nil
+}
+
+// FunctionName implements sql.FunctionExpression
+func (s *SRID) FunctionName() string {
+	return "st_srid"
+}
+
+// Description implements sql.FunctionExpression
+func (s *SRID) Description() string {
+	return "returns the SRID value of given geometry object. If given a second argument, returns a new geometry object with second argument as SRID value."
+}
+
+// Type implements the sql.Expression interface.
+func (s *SRID) Type() sql.Type {
+	if len(s.ChildExpressions) == 1 {
+		return sql.Int32
+	} else {
+		return s.ChildExpressions[0].Type()
+	}
+}
+
+func (s *SRID) String() string {
+	var args = make([]string, len(s.ChildExpressions))
+	for i, arg := range s.ChildExpressions {
+		args[i] = arg.String()
+	}
+	return fmt.Sprintf("ST_SRID(%s)", strings.Join(args, ","))
+}
+
+// WithChildren implements the Expression interface.
+func (s *SRID) WithChildren(children ...sql.Expression) (sql.Expression, error) {
+	return NewSRID(children...)
+}
+
+// DeepCopyPointSRID creates a deep copy of point object with given SRID
+func DeepCopyPointSRID(p sql.Point, srid uint32) sql.Point {
```

Comment:
For things like this, we usually define something like WithSRID for these types, and it returns a copy with the SRID given

Reply:
Generally speaking we want each expression type to own its internal logic, including constructing structs of itself